From f8ddf4dd6cbdfd896864c08f8ec359af85e4ab3c Mon Sep 17 00:00:00 2001
From: callen <callen.cai@rock-chips.com>
Date: Mon, 5 Jun 2017 14:48:29 +0800
Subject: [PATCH 13/14] StorageManager rk3368H/rk3399 read calculated size in
 "/sys/block/mmcblk1/size". This is verify on
 rk3368,rk3399.Bug replay step: 1.Check origin storage
 size in Settings; 2.Insert a TF card to device(TF
 card size is less than internal storage will be
 obviously); 3.Check storage again,it will show a
 wrong internal storage size.

Change-Id: Id53081e486dc398e2bfc9c7af82f69f2d25b02e8

Signed-off-by: tug <tug@skyth-tek.com>
---
 core/java/android/os/storage/StorageManager.java |   25 ++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/core/java/android/os/storage/StorageManager.java b/core/java/android/os/storage/StorageManager.java
index c5507b9..7c47b82 100644
--- a/core/java/android/os/storage/StorageManager.java
+++ b/core/java/android/os/storage/StorageManager.java
@@ -59,6 +59,7 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 import java.util.concurrent.atomic.AtomicInteger;
+import android.os.SystemProperties;
 
 /**
  * StorageManager is the interface to the systems storage service. The storage
@@ -140,6 +141,12 @@ public class StorageManager {
             "/sys/block/mmcblk0/size",
             "/sys/block/sda/size"
     };
+    private static final String[] INTERNAL_STORAGE_SIZE_PATHS_ALTERNATIVE = {
+            "/sys/block/mmcblk1/size",
+            "/sys/block/sda/size"
+    };
+    private static String platform=SystemProperties.get("ro.board.platform", "unknow");
+    private static boolean alternative_path=platform.equals("rk3368")||platform.equals("rk3399");
     private static final int INTERNAL_STORAGE_SECTOR_SIZE = 512;
 
     private final Context mContext;
@@ -931,10 +938,20 @@ public class StorageManager {
 
     /** {@hide} */
     public long getPrimaryStorageSize() {
-        for (String path : INTERNAL_STORAGE_SIZE_PATHS) {
-            final long numberBlocks = readLong(path);
-            if (numberBlocks > 0) {
-                return numberBlocks * INTERNAL_STORAGE_SECTOR_SIZE;
+        if(alternative_path){
+            for (String path : INTERNAL_STORAGE_SIZE_PATHS_ALTERNATIVE) {
+                final long numberBlocks = readLong(path);
+                if (numberBlocks > 0) {
+                    return numberBlocks * INTERNAL_STORAGE_SECTOR_SIZE;
+                }
+            }
+        }
+        else{
+            for (String path : INTERNAL_STORAGE_SIZE_PATHS) {
+                final long numberBlocks = readLong(path);
+                if (numberBlocks > 0) {
+                    return numberBlocks * INTERNAL_STORAGE_SECTOR_SIZE;
+                }
             }
         }
         return 0;
-- 
1.7.9.5

